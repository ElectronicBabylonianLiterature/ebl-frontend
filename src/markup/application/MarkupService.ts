import Bluebird from 'bluebird'
import ApiClient from 'http/ApiClient'
import ReferenceInjector from 'transliteration/application/ReferenceInjector'
import BibliographyService from 'bibliography/application/BibliographyService'
import { MarkupPart } from 'transliteration/domain/markup'
import { stringify } from 'query-string'

export default class MarkupService {
  private readonly referenceInjector: ReferenceInjector
  constructor(
    private readonly apiClient: ApiClient,
    bibliographyService: BibliographyService
  ) {
    this.referenceInjector = new ReferenceInjector(bibliographyService)
  }

  fromString(text: string): Bluebird<readonly MarkupPart[]> {
    return this.apiClient
      .fetchJson(
        `/markup?${stringify({
          text: text,
        })}`,
        false
      )
      .then((parts) => {
        return Bluebird.all(
          parts && Bluebird.all(this.injectReferencesToMarkup(parts))
        )
      })
  }

  injectReferencesToMarkup(
    parts: readonly MarkupPart[]
  ): Bluebird<readonly MarkupPart[]> {
    return this.referenceInjector.injectReferencesToMarkup(parts)
  }
}
