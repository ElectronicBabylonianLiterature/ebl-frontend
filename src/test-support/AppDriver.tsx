import React from 'react'
import {
  fireEvent,
  render,
  RenderResult,
  act,
  Matcher,
  within,
  waitFor,
} from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import _ from 'lodash'
import App from 'App'
import WordRepository from 'dictionary/infrastructure/WordRepository'
import FragmentRepository from 'fragmentarium/infrastructure/FragmentRepository'
import ApiImageRepository from 'fragmentarium/infrastructure/ImageRepository'
import FragmentService from 'fragmentarium/application/FragmentService'
import WordService from 'dictionary/application/WordService'
import TextService from 'corpus/application/TextService'
import MemorySession, { Session, guestSession } from 'auth/Session'
import BibliographyRepository from 'bibliography/infrastructure/BibliographyRepository'
import BibliographyService from 'bibliography/application/BibliographyService'
import FragmentSearchService from 'fragmentarium/application/FragmentSearchService'
import { Promise } from 'bluebird'
import { submitForm } from 'test-support/utils'
import { eblNameProperty, AuthenticationContext } from 'auth/Auth'
import SignRepository from 'signs/infrastructure/SignRepository'
import SignService from 'signs/application/SignService'

function createApp(api): JSX.Element {
  const wordRepository = new WordRepository(api)
  const fragmentRepository = new FragmentRepository(api)
  const imageRepository = new ApiImageRepository(api)
  const bibliographyRepository = new BibliographyRepository(api)
  const wordService = new WordService(wordRepository)
  const bibliographyService = new BibliographyService(bibliographyRepository)
  const fragmentService = new FragmentService(
    fragmentRepository,
    imageRepository,
    wordRepository,
    bibliographyService
  )
  const fragmentSearchService = new FragmentSearchService(fragmentRepository)
  const textService = new TextService(api, fragmentService, wordService)
  const signsRepository = new SignRepository(api)
  const signService = new SignService(signsRepository)
  return (
    <App
      signService={signService}
      wordService={wordService}
      fragmentService={fragmentService}
      fragmentSearchService={fragmentSearchService}
      bibliographyService={bibliographyService}
      textService={textService}
    />
  )
}

export default class AppDriver {
  private readonly api
  private initialEntries: string[] = []
  private view: RenderResult | null = null
  private session: Session | null = null

  constructor(api) {
    this.api = api
  }

  getView(): RenderResult {
    if (this.view) {
      return this.view
    } else {
      throw new Error('getElement called before render.')
    }
  }

  withPath(path: string): AppDriver {
    this.initialEntries = [path]
    return this
  }

  withSession(): AppDriver {
    this.session = new MemorySession([
      'write:texts',
      'read:fragments',
      'annotate:fragments',
      'read:words',
    ])
    return this
  }

  async render(): Promise<AppDriver> {
    await act(async () => {
      this.view = render(
        <MemoryRouter initialEntries={this.initialEntries}>
          <AuthenticationContext.Provider
            value={{
              login: _.noop,
              logout: _.noop,
              getSession: (): Session => this.session ?? guestSession,
              isAuthenticated: (): boolean => this.session !== null,
              getAccessToken(): Promise<string> {
                throw new Error('Not implemented')
              },
              getUser(): { [eblNameProperty]: string } {
                return { [eblNameProperty]: 'Test' }
              },
            }}
          >
            {createApp(this.api)}
          </AuthenticationContext.Provider>
        </MemoryRouter>
      )
    })

    return this
  }

  async waitForText(text: Matcher): Promise<void> {
    await this.getView().findByText(text)
  }

  async waitForTextToDisappear(text: Matcher): Promise<void> {
    await waitFor(() => {
      expect(this.getView().queryByText(text)).not.toBeInTheDocument()
    })
  }

  expectTextContent(text: string | RegExp): void {
    expect(this.getView().container).toHaveTextContent(text)
  }

  expectNotInContent(text: Matcher): void {
    expect(this.getView().queryByText(text)).not.toBeInTheDocument()
  }

  expectLink(text: Matcher, expectedHref: string): void {
    expect(this.getView().getByText(text)).toHaveAttribute('href', expectedHref)
  }

  expectBreadcrumbs(crumbs: readonly string[]): void {
    this.expectTextContent(crumbs.join(''))
  }

  expectBreadcrumb(crumb: string, link: string): void {
    expect(
      within(this.getView().getByLabelText('breadcrumb')).getByText(crumb)
    ).toHaveAttribute('href', link)
  }

  expectInputElement(label: Matcher, expectedValue: unknown): void {
    expect(this.getView().getByLabelText(label)).toHaveValue(
      String(expectedValue)
    )
  }

  expectChecked(label: Matcher): void {
    expect(this.getView().getByLabelText(label)).toBeChecked()
  }

  expectNotChecked(label: Matcher): void {
    expect(this.getView().getByLabelText(label)).not.toBeChecked()
  }

  async changeValueByLabel(label: Matcher, newValue: unknown): Promise<void> {
    const input = this.getView().getByLabelText(label)
    await act(async () => {
      fireEvent.change(input, { target: { value: newValue } })
    })
  }

  async submitForm(): Promise<void> {
    await submitForm(this.getView().container)
  }

  async click(text: Matcher, n = 0): Promise<void> {
    const clickable = this.getView().getAllByText(text)[n]
    await act(async () => {
      fireEvent.click(clickable)
    })
  }
}
