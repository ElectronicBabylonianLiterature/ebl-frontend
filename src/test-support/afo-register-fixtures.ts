import { Factory } from 'fishery'
import Chance from 'chance'
import AfoRegisterRecord, {
  AfoRegisterRecordSuggestion,
} from 'afo-register/domain/Record'

const chance = new Chance()
const PUBLICATIONS = [
  'StOr',
  'Al.T.',
  'OECT',
  'VS',
  'STT',
  'CCT',
  'CM',
  'CST',
  'SAAB',
]

const getRandomPublication = (): string => chance.pickone(PUBLICATIONS)

const getAfoNumber = (): string => `AfO ${chance.integer({ min: 10, max: 52 })}`

const getPage = (): string => `${chance.integer({ min: 1, max: 700 })}`

const getText = (): string =>
  `${getRandomPublication()}, ${chance.integer({ min: 1, max: 40 })}`

const getTextNumber = (): string =>
  `Nr. ${chance.integer({ min: 1, max: 300 })}`

const getTextNumbers = (): string[] =>
  Array.from({ length: chance.integer({ min: 1, max: 15 }) }, getTextNumber)

const getLinesDiscussed = (): string =>
  `${chance.integer({ min: 1, max: 40 })}f.`

const getDiscussedBy = (): string =>
  `${chance.last()}, ${getRandomPublication()}, ${chance.integer({
    min: 1,
    max: 40,
  })}, ${chance.integer({ min: 1, max: 50 })}`

const getFragmentNumber = (): string[] =>
  Array.from(
    { length: chance.integer({ min: 0, max: 5 }) },
    () => `${chance.word()}.${chance.natural()}`
  )

export const afoRegisterRecordFactory = Factory.define<AfoRegisterRecord>(
  () =>
    new AfoRegisterRecord({
      afoNumber: getAfoNumber(),
      page: getPage(),
      text: getText(),
      textNumber: getTextNumber(),
      linesDiscussed: getLinesDiscussed(),
      discussedBy: getDiscussedBy(),
      discussedByNotes: chance.sentence(),
      fragmentNumbers: getFragmentNumber(),
    })
)

export const afoRegisterRecordSuggestionFactory = Factory.define<
  AfoRegisterRecordSuggestion
>(
  () =>
    new AfoRegisterRecordSuggestion({
      text: getText(),
      textNumbers: getTextNumbers().sort(),
    })
)
