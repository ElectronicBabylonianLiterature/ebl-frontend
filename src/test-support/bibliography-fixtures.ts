import { Factory } from 'fishery'
import Chance from 'chance'
import Reference, { ReferenceType } from 'bibliography/domain/Reference'
import BibliographyEntry, {
  CslData,
} from 'bibliography/domain/BibliographyEntry'
import { ReferenceDto } from 'bibliography/domain/referenceDto'

const defaultChance = new Chance()

function integer(min: number, max: number, chance?: Chance.Chance): number {
  return (chance || defaultChance).integer({ min: min, max: max })
}

function type(chance = defaultChance): ReferenceType {
  return chance.pickone([
    'EDITION',
    'DISCUSSION',
    'COPY',
    'PHOTO',
    'TRANSLATION',
  ])
}

const authorFactory = Factory.define<
  { given: string; family: string },
  { chance: Chance.Chance }
>(({ transientParams }) => {
  const chance = transientParams.chance ?? defaultChance
  return {
    given: chance.first(),
    family: chance.last(),
  }
})

export const cslDataFactory = Factory.define<
  CslData,
  { chance: Chance.Chance }
>(({ transientParams }) => {
  const chance = transientParams.chance ?? defaultChance
  const issuedDate = chance.date()
  return {
    id: chance.guid(),
    title: chance.sentence(),
    type: chance.pickone(['article-journal', 'paper-conference']),
    issued: {
      'date-parts': [
        [issuedDate.getFullYear(), issuedDate.getMonth(), issuedDate.getDate()],
      ],
    },
    volume: integer(1, 99, chance),
    page: `${integer(1, 99, chance)}-${integer(100, 999, chance)}`,
    issue: integer(1, 99, chance),
    'container-title': chance.sentence(),
    author: authorFactory.buildList(2, {}, { transient: { chance: chance } }),
    URL: chance.url(),
  }
})

export const cslDataWithContainerTitleShortFactory = cslDataFactory.params({
  'container-title-short': defaultChance.syllable(),
})

export const bibliographyEntryFactory = Factory.define<
  BibliographyEntry,
  CslData & { chance: Chance.Chance }
>(({ transientParams }) => {
  const chance = transientParams.chance ?? defaultChance
  return new BibliographyEntry(
    cslDataFactory.build(transientParams, { transient: { chance } })
  )
})

export function buildBorger1957(): BibliographyEntry {
  return bibliographyEntryFactory.build(
    {},
    {
      transient: {
        author: [{ family: 'Borger' }],
        issued: { 'date-parts': [[1957]] },
      },
    }
  )
}

export const referenceDtoFactory = Factory.define<
  ReferenceDto,
  { chance: Chance.Chance }
>(({ associations, transientParams }) => {
  const chance = transientParams.chance ?? defaultChance
  return {
    id: chance.string(),
    type: type(chance),
    pages: `${chance.natural()}-${chance.natural()}`,
    notes: chance.sentence(),
    linesCited: chance.pickset(['1.', '2.', "3'.", "4'.2."], 2),
    document: associations.document ?? null,
  }
})

export const referenceFactory = Factory.define<
  Reference,
  { chance: Chance.Chance }
>(({ associations, transientParams }) => {
  const chance = transientParams.chance ?? defaultChance

  return new Reference(
    type(chance),
    `${chance.natural()}-${chance.natural()}`,
    chance.sentence(),
    chance.pickset(['1.', '2.', "3'.", "4'.2."], 2),
    associations.document ??
      bibliographyEntryFactory.build({}, { transient: { chance } })
  )
})

export function buildReferenceWithContainerTitle(
  type: ReferenceType,
  cslData = {}
): Reference {
  return referenceFactory.build({
    type: type,
    document: bibliographyEntryFactory.build(
      {},
      { transient: cslDataWithContainerTitleShortFactory.build(cslData) }
    ),
  })
}

export function buildReferenceWithManyAuthors(): Reference {
  return referenceFactory.build({
    type: 'COPY',
    document: bibliographyEntryFactory.build(
      {},
      { transient: { author: authorFactory.buildList(4) } }
    ),
  })
}
