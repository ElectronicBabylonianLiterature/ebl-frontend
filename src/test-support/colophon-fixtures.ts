import { Factory } from 'fishery'
import { Chance } from 'chance'
import {
  IndividualType,
  ColophonStatus,
  ColophonOwnership,
  ColophonType,
  NameAttestation,
  IndividualAttestation,
  Colophon,
  ProvenanceAttestation,
  IndividualTypeAttestation,
} from 'fragmentarium/domain/Colophon'
import { Provenances } from 'corpus/domain/provenance'

const chance = new Chance()

const nameAttestationFactory = Factory.define<NameAttestation>(() => ({
  value: chance.name(),
  isBroken: chance.bool(),
  isUncertain: chance.bool(),
}))

const provenanceAttestationFactory = Factory.define<ProvenanceAttestation>(
  () => ({
    value: chance.pickone(Object.values(Provenances)).name,
    isBroken: chance.bool(),
    isUncertain: chance.bool(),
  })
)

const individualTypeAttestationFactory = Factory.define<
  IndividualTypeAttestation
>(() => ({
  value: chance.pickone(Object.values(IndividualType)),
  isBroken: chance.bool(),
  isUncertain: chance.bool(),
}))

const individualAttestationFactory = Factory.define<IndividualAttestation>(
  () => {
    return new IndividualAttestation({
      name: nameAttestationFactory.build(),
      sonOf: chance.bool() ? nameAttestationFactory.build() : undefined,
      grandsonOf: chance.bool() ? nameAttestationFactory.build() : undefined,
      family: chance.bool() ? nameAttestationFactory.build() : undefined,
      nativeOf: provenanceAttestationFactory.build(),
      type: individualTypeAttestationFactory.build(),
    })
  }
)

export const colophonFactory = Factory.define<Colophon>(() => ({
  colophonOwnership: chance.pickone(Object.values(ColophonOwnership)),
  colophonType: chance.pickone(Object.values(ColophonType)),
  colophonStatus: chance.pickone(Object.values(ColophonStatus)),
  originalFrom: provenanceAttestationFactory.build(),
  writtenIn: provenanceAttestationFactory.build(),
  notesToScribalProcess: chance.sentence(),
  individuals: individualAttestationFactory.buildList(
    chance.integer({ min: 1, max: 5 })
  ),
}))
