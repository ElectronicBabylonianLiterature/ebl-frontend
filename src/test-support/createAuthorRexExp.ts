import _ from 'lodash'
import BibliographyEntry from 'bibliography/domain/BibliographyEntry'

export default function createAuthorRegExp(entry: BibliographyEntry): RegExp {
  return new RegExp(_.escapeRegExp(entry.primaryAuthor.replace(/'/g, 'â€™')))
}
