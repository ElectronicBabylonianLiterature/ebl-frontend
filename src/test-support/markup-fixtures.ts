import Reference from 'bibliography/domain/Reference'
import { MarkupPart } from 'transliteration/domain/markup'
import BibliographyEntry from 'bibliography/domain/BibliographyEntry'

export const markupString = `This tablet is a one columned chronicle-fragment, telling about the faulty reignship of king Šulgi, who committed sins against Babylon and Uruk. The text is written in an accusatory tone, stressed by the repetition of exclamatory sentences about Šulgis sinfull deeds. It was discussed in lenghth by @bib{RN891@63-72}, who pointed out its inspiration trough the Sumerian Kinglist as well as anachronistic allusions to Nabonid.
The tablet is part of a series, as can be seen from the existence of the catchline and a “specular catchline” as it is called by Hunger, (@i{SpTU} 1, 20 n. 2), that seems to resume the content of the preceding chapter. About one half or even two thirds of the composition is missing. This is underlined by the colophon, that takes almost all of the space on the reverse but in many other cases covers only about a third and occasionally half of a tablet.
The tablet stems from the 27. campaign in Uruk 1969 of the residential area U XVIII and was published first by Hunger 1976 in SpTU 1, 2.`

const markupPartOne: MarkupPart = {
  text:
    'This tablet is a one columned chronicle-fragment, telling about the faulty reignship of king Šulgi, who committed sins against Babylon and Uruk. The text is written in an accusatory tone, stressed by the repetition of exclamatory sentences about Šulgis sinfull deeds. It was discussed in lenghth by ',
  type: 'StringPart',
}

const markupPartTwoSerialized: MarkupPart = {
  reference: new Reference(
    'DISCUSSION',
    '63-72',
    '',
    [],
    new BibliographyEntry({
      cslData: {
        id: 'RN891',
        author: [
          {
            family: 'Cavigneaux',
            given: 'A.',
          },
        ],
        title: 'Shulgi, Nabonide et les grecs',
        'container-title':
          'An Experienced Scribe who Neglects Nothing. Ancient Near Eastern Studies in Honor of Jacob Klein',
        editor: [
          {
            given: 'Y.',
            family: 'Sefati',
          },
          {
            given: 'P.',
            family: 'Artzi',
          },
          {
            given: 'Ch.',
            family: 'Cohen',
          },
          {
            given: 'B.L.',
            family: 'Eichler',
          },
          {
            given: 'V.A.',
            family: 'Hurowitz',
          },
        ],
        publisher: 'CDL',
        'publisher-place': 'Bethesda',
        page: '63-72',
        type: 'chapter',
        'citation-label': 'RN891',
        issued: {
          'date-parts': [[2005]],
        },
      },
    })
  ),
  type: 'BibliographyPart',
}

const markupPartThree: MarkupPart = {
  text:
    ', who pointed out its inspiration trough the Sumerian Kinglist as well as anachronistic allusions to Nabonid. The tablet is part of a series, as can be seen from the existence of the catchline and a “specular catchline” as it is called by Hunger, (',
  type: 'StringPart',
}

const markupPartFour: MarkupPart = {
  text: 'SpTU',
  type: 'EmphasisPart',
}

const markupPartFive: MarkupPart = {
  text:
    ' 1, 20 n. 2), that seems to resume the content of the preceding chapter. About one half or even two thirds of the composition is missing. This is underlined by the colophon, that takes almost all of the space on the reverse but in many other cases covers only about a third and occasionally half of a tablet. The tablet stems from the 27. campaign in Uruk 1969 of the residential area U XVIII and was published first by Hunger 1976 in SpTU 1, 2.',
  type: 'StringPart',
}

export const markupDtoSerialized = [
  markupPartOne,
  markupPartTwoSerialized,
  markupPartThree,
  markupPartFour,
  markupPartFive,
]
