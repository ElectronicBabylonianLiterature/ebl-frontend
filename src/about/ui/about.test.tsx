import React from 'react'
import About from 'about/ui/about'
import Bluebird from 'bluebird'
import '@testing-library/jest-dom/extend-expect'
import MarkupService from 'markup/application/MarkupService'
import { markupDtoSerialized } from 'test-support/markup-fixtures'
import { act, render } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'

jest.mock('markup/application/MarkupService')

let mockDate: jest.SpyInstance<string, any>

beforeAll(() => {
  mockDate = jest.spyOn(Date.prototype, 'toLocaleDateString')
})

afterAll(() => {
  mockDate.mockRestore()
})

const markupServiceMock = new (MarkupService as jest.Mock<
  jest.Mocked<MarkupService>
>)()

test('Snapshot', async () => {
  mockDate.mockReturnValue('1/1/2023')
  markupServiceMock.fromString.mockReturnValue(
    Bluebird.resolve(markupDtoSerialized)
  )
  let container
  await act(async () => {
    container = await render(
      <MemoryRouter>
        <About markupService={markupServiceMock} activeTab="corpus" />
      </MemoryRouter>
    ).container
  })
  expect(container).toMatchSnapshot()
})
