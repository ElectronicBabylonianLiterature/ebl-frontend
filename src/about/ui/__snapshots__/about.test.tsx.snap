// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<div>
  <main
    class="main"
  >
    <div
      class="main__content"
    >
      <header
        class="main__header"
      >
        <nav
          aria-label="breadcrumb"
          class="main__breadcrumbs"
        >
          <ol
            class="breadcrumb"
          >
            <li
              class="breadcrumb-item active"
            >
              <a
                href="/"
              >
                eBL
              </a>
            </li>
            <li
              aria-current="page"
              class="breadcrumb-item active"
            >
              About
            </li>
            <li
              aria-current="page"
              class="breadcrumb-item active"
            >
              Corpus
            </li>
          </ol>
        </nav>
        <h2
          class="main__heading"
        >
          About
        </h2>
        <div
          class="main__toolbar btn-toolbar"
          role="toolbar"
        />
      </header>
      <nav
        class="nav nav-tabs"
        role="tablist"
      >
        <a
          aria-controls="about-tabpane-project"
          aria-selected="false"
          class="nav-item nav-link"
          data-rb-event-key="project"
          href="#"
          id="about-tab-project"
          role="tab"
        >
          eBL Project
        </a>
        <a
          aria-controls="about-tabpane-library"
          aria-selected="false"
          class="nav-item nav-link"
          data-rb-event-key="library"
          href="#"
          id="about-tab-library"
          role="tab"
        >
          Library
        </a>
        <a
          aria-controls="about-tabpane-corpus"
          aria-selected="true"
          class="nav-item nav-link active"
          data-rb-event-key="corpus"
          href="#"
          id="about-tab-corpus"
          role="tab"
        >
          Corpus
        </a>
        <a
          aria-controls="about-tabpane-signs"
          aria-selected="false"
          class="nav-item nav-link"
          data-rb-event-key="signs"
          href="#"
          id="about-tab-signs"
          role="tab"
        >
          Signs
        </a>
        <a
          aria-controls="about-tabpane-dictionary"
          aria-selected="false"
          class="nav-item nav-link"
          data-rb-event-key="dictionary"
          href="#"
          id="about-tab-dictionary"
          role="tab"
        >
          Dictionary
        </a>
        <a
          aria-controls="about-tabpane-bibliography"
          aria-selected="false"
          class="nav-item nav-link"
          data-rb-event-key="bibliography"
          href="#"
          id="about-tab-bibliography"
          role="tab"
        >
          Bibliography
        </a>
        <a
          aria-controls="about-tabpane-news"
          aria-selected="false"
          class="nav-item nav-link"
          data-rb-event-key="news"
          href="#"
          id="about-tab-news"
          role="tab"
        >
          News
        </a>
      </nav>
      <div
        class="tab-content"
      >
        <div
          aria-hidden="false"
          aria-labelledby="about-tab-corpus"
          class="fade tab-pane active show"
          id="about-tabpane-corpus"
          role="tabpanel"
        >
          <blockquote>
            <p>
              <em>
                Ohne feste kritische Grundlage wird das philologische Gebäude auf Sand aufgeführt und die philologische Wissenschaft gestaltet sich zum bloßen Dilettantismus.
              </em>
            </p>
            <p>
              W. Freund, 
              <a
                href="https://books.google.de/books?id=PFJeAAAAcAAJ&printsec=frontcover&source=gbs_ge_summary_r&cad=0#v=onepage&q&f=false"
              >
                <em>
                  Triennium philologicum oder Grundzüge der philologischen Wissenschaften
                </em>
              </a>
              . Leipzig, 1874.
            </p>
          </blockquote>
          <p>
            The eBL editions aim to present the best text that can be reconstructed at  present. The editions prepared in the course of the project include all  previous scholarship on the texts, and in particular all new manuscripts  identified after the last printed editions. The eBL edition of the 
            <a
              href="/corpus/L/1/12"
            >
              <em>
                Cuthaean  Legend of Narām-Sîn
              </em>
            </a>
            , for instance, almost doubles the  manuscript basis of the text available to its last editor; that of the  
            <a
              href="/corpus/L/2/3"
            >
              <em>
                Counsels of Wisdom
              </em>
            </a>
             includes over twenty new manuscripts that  were absent from the most recent printed edition. Most of the new manuscripts  used have been identified by the eBL team, and are currently being published in  the series of articles 
            <em>
              From the Electronic Babylonian Literature Lab
            </em>
             that  appear in the journal 
            <em>
              Kaskal
            </em>
            . Moreover, the eBL editions are constantly  updated, and the editors will incorporate new discoveries as they appear.
          </p>
          <h3>
            I. Corpus
          </h3>
          <p>
            <span>
              This tablet is a one columned chronicle-fragment, telling about the faulty reignship of king Šulgi, who committed sins against Babylon and Uruk. The text is written in an accusatory tone, stressed by the repetition of exclamatory sentences about Šulgis sinfull deeds. It was discussed in lenghth by 
            </span>
            <span
              class="reference-popover__citation"
            >
              , : 63-72
              <span
                class="type-abbreviation"
              >
                 (D)
              </span>
            </span>
            <span>
              , who pointed out its inspiration trough the Sumerian Kinglist as well as anachronistic allusions to Nabonid. The tablet is part of a series, as can be seen from the existence of the catchline and a “specular catchline” as it is called by Hunger, (
            </span>
            <em>
              SpTU
            </em>
            <span>
               1, 20 n. 2), that seems to resume the content of the preceding chapter. About one half or even two thirds of the composition is missing. This is underlined by the colophon, that takes almost all of the space on the reverse but in many other cases covers only about a third and occasionally half of a tablet. The tablet stems from the 27. campaign in Uruk 1969 of the residential area U XVIII and was published first by Hunger 1976 in SpTU 1, 2.
            </span>
          </p>
          <h3>
            II. Translations
          </h3>
          <figure
            class="Introduction__photoRight"
          >
            <a
              href="https://www.eisenbrauns.org/books/titles/978-1-883053-76-5.html"
            >
              <img
                alt="Foster, Before the Muses"
                class="Introduction__200px"
                src="muses.jpg"
              />
            </a>
            <figcaption
              class="Introduction__caption"
            >
              Foster, 
              <em>
                Before the Muses
              </em>
            </figcaption>
          </figure>
          <p>
            <span>
              This tablet is a one columned chronicle-fragment, telling about the faulty reignship of king Šulgi, who committed sins against Babylon and Uruk. The text is written in an accusatory tone, stressed by the repetition of exclamatory sentences about Šulgis sinfull deeds. It was discussed in lenghth by 
            </span>
            <span
              class="reference-popover__citation"
            >
              , : 63-72
              <span
                class="type-abbreviation"
              >
                 (D)
              </span>
            </span>
            <span>
              , who pointed out its inspiration trough the Sumerian Kinglist as well as anachronistic allusions to Nabonid. The tablet is part of a series, as can be seen from the existence of the catchline and a “specular catchline” as it is called by Hunger, (
            </span>
            <em>
              SpTU
            </em>
            <span>
               1, 20 n. 2), that seems to resume the content of the preceding chapter. About one half or even two thirds of the composition is missing. This is underlined by the colophon, that takes almost all of the space on the reverse but in many other cases covers only about a third and occasionally half of a tablet. The tablet stems from the 27. campaign in Uruk 1969 of the residential area U XVIII and was published first by Hunger 1976 in SpTU 1, 2.
            </span>
          </p>
          <p>
            Additional translations have been produced by A. R. George (
            <em>
              Gilgameš
            </em>
            ), W. G.  Lambert (
            <a
              href="/corpus/L/3/3"
            >
              A Syncretistic Hymn to Ištar
            </a>
             and 
            <a
              href="/corpus/L/3/3"
            >
              Marduk’s Address to  the Demons
            </a>
            ), T. Mitto (
            <a
              href="/corpus/L/0/0"
            >
              Catalogue of Texts and  Authors
            </a>
             and 
            <a
              href="/corpus/L/3/10"
            >
              Hymn to Ninurta as Savior
            </a>
            ) and E.  Jiménez.
          </p>
          <p>
            In addition, a series of translations into Arabic are in preparation by A. A.  Fadhil, W. Khatabe, and W. Zerkly (see for now A. A. Fadhil’s Arabic  translation of 
            <a
              href="/corpus/L/1/2/SB/I"
            >
              <em>
                Enūma eliš
              </em>
               I
            </a>
            )
          </p>
          <h3>
            III. Ideal Text
          </h3>
          <p>
            The main version displayed on the eBL Corpus is a phonetic transcription of the  text, which has been adjusted according to the rules of Standard Babylonian  grammar. This practice somewhat departs from the Assyriological tradition of  editing “eclectic” texts, i.e. transliterations that combine the readings of  various manuscripts. It has been adopted, however, in the conviction that  Mesopotamian texts are also objects of art, and not just objects of scientific  study, and as such convey their message only through the interplay of form and  content.
          </p>
          <figure
            class="Introduction__photoLeft"
          >
            <img
              alt="eBL edition of Enūma eliš I 49"
              class="Introduction__450px"
              src="ee49.png"
            />
            <figcaption
              class="Introduction__caption"
            >
              eBL edition of 
              <em>
                Enūma eliš
              </em>
               
              <a
                href="/corpus/L/1/2/SB/I#49"
              >
                I 49
              </a>
            </figcaption>
          </figure>
          <p>
            The practice of using a phonetic transcription as the main text no doubt has  disadvantages: for instance, it obscures the way in which the text is written  in cuneiform, and it affords a sense of grammatical certainty that is absent  from a regular transliteration. However, it also offers considerable  advantages: in particular, it does not require the editor to adopt any  particular spelling when no good criteria exist for preferring one over the  other. In 
            <em>
              Enūma eliš
            </em>
             
            <a
              href="/corpus/L/1/2/SB/I#49"
            >
              I 49
            </a>
             (see the adjoining image),  for instance, the editor would have to choose between the accusative  
            <em>
              al-ka-ta
            </em>
            , attested only in Assyrian manuscripts, or the normal Babylonian  spelling 
            <em>
              al-ka-tu
            </em>
            ₄. In a transcription the editor can convey his  interpretation of the text in a much more satisfactory manner than in a  traditional transliteration.
          </p>
          <p>
            The transcription respects the ways in which the manuscripts are written as  much as possible. For instance, 
            <em>
              Enūma eliš
            </em>
             
            <a
              href="/corpus/L/1/2/SB/VI#124"
            >
              VI 124
            </a>
              is transcribed as 
            <em>
              muṭaḫḫidu urîšun
            </em>
            , “who enriches their stables,” and thus  assumes a hymno-epic ending -
            <em>
              u
            </em>
             of the 
            <em>
              nomen regens
            </em>
            , instead of the normal  bound form 
            <em>
              muṭaḫḫid
            </em>
            , since 
            <em>
              mu-ṭaḫ-ḫi-du
            </em>
             is the spelling of all manuscripts.
          </p>
          <h3>
            IV. Score Edition
          </h3>
          <figure
            class="Introduction__photoRight"
          >
            <img
              alt="eBL edition of Šamaš Hymn 134"
              class="Introduction__450px"
              src="shamash134.jpg"
            />
            <figcaption
              class="Introduction__caption"
            >
              eBL edition of 
              <em>
                Šamaš Hymn
              </em>
               
              <a
                href="/corpus/L/3/4/SB/-#134"
              >
                134
              </a>
            </figcaption>
          </figure>
          <p>
            Since the number of manuscripts of each text is constantly growing, cuneiform  studies is reaching the point where it is no longer possible to print text  editions in the score format. Just the eBL edition score of 
            <em>
              Enūma eliš
            </em>
            , for  example, would require some 300 pages in font size 10. Despite this technical  limitation, scores are the fastest, most straightforward way of checking  exactly how manuscripts write their texts. The eBL scores are, moreover,  aligned with the ideal line, so that the reader can check the manuscript basis  of the editor’s decisions at any time.
          </p>
          <p>
            All transliterations have been checked twice against the published copies and  against photographs and in some cases the originals of the cuneiform tablets.
          </p>
          <h3>
            V. Paratextual information
          </h3>
          <p>
            Each text is furnished with an introduction, which discusses the content and  structure of the text, its origins and transmission, its Sitz im Leben and the  history of research concerning the text.
          </p>
          <p>
            The editions are fully lemmatized and annotated. The annotation includes  indications of all parallel lines, so that the text can be studied as part of  the intertextual network in which it belongs. In addition, the notes on  individual lines endeavor to provide the reader with references to all previous  bibliography, with particular emphasis on studies which have appeared in the  last few years. The notes on the score edition discuss mostly philological  issues pertaining to an individual manuscript.
          </p>
          <p>
            The colophons of the individual manuscripts are transliterated independently,  and can be accessed on the homepage of any text (e.g. 
            <a
              href="/corpus/L/1/4"
            >
              here
            </a>
            ).  In some cases, manuscripts include lines that, though clearly part of the  composition in question, cannot yet be placed in it, and thus are  transliterated independently, as Unplaced lines.
          </p>
          <h3>
            VI. Šumma ālu
          </h3>
          <p>
            The editions of 
            <em>
              Šumma ālu
            </em>
             presented here were prepared in the context of the  projects “Edition of the Omen Series 
            <em>
             Šumma ālu
            </em>
             ” (2017–2021;  
            <a
              href="http://p3.snf.ch/project-175970"
            >
              http://p3.snf.ch/project-175970
            </a>
            ) and  “Typology and potential of the excerpt tablets of 
            <em>
            Šumma ālu
            </em>
            ” (2022–2023;  
            <a
              href="http://p3.snf.ch/project-205122"
            >
              http://p3.snf.ch/project-205122
            </a>
            ), both  directed by Prof. Catherine Mittermayer at the University of Geneva and funded  by the Swiss National Science Foundation. The complete score editions can be  downloaded (PDF) at the “Archive ouverte” of the University of Geneva  (
            <a
              href="https://archive-ouverte.unige.ch/"
            >
              https://archive-ouverte.unige.ch/
            </a>
            ; search  for “Shumma alu”).
          </p>
          <figure
            class="Introduction__photoCentered"
          >
            <a
              href="https://www.unige.ch/"
            >
              <img
                alt="Université de Genève"
                src="geneva.jpg"
              />
            </a>
            <a
              href="https://www.snf.ch/fr"
            >
              <img
                alt="Fonds national suisse de la recherche scientifique"
                src="snf.png"
              />
            </a>
          </figure>
        </div>
      </div>
    </div>
  </main>
</div>
`;
